CDTDatastore/HTTP/CDTIAMSessionCookieInterceptor.m

```diff
@@ -0,0 +1,145 @@
+#import <Foundation/Foundation.h>
+
+#import "CDTIAMSessionCookieInterceptor.h"
+#import "CDTLogging.h"
+
+static const NSInteger CDTIAMSessionCookieRequestTimeout = 600;
+
+
+@interface CDTIAMSessionCookieInterceptor ()
+
+
+@property NSData *iamSessionRequestBody;
+
+- (NSData *)getBearerToken;
+
+/** NSURLSession to make calls to get IAM bearer token (shouldn't be same one we're intercepting). */
+@property (nonnull, nonatomic, strong) NSURLSession *iamUrlSession;
+
+@property NSURL *iamTokenUrl;
+
+@end
+
+@implementation CDTIAMSessionCookieInterceptor
+
+- (instancetype)initWithAPIKey:(NSString *)apiKey
+{
+    self = [super init];
+    if (self) {
+        // build the request to get the IAM bearer token
+        _iamSessionRequestBody = [[NSString stringWithFormat:@"grant_type=urn:ibm:params:oauth:grant-type:apikey&response_type=cloud_iam&apikey=%@", apiKey] dataUsingEncoding:NSUTF8StringEncoding];
+        
+        NSURLSessionConfiguration *config =
+        [NSURLSessionConfiguration ephemeralSessionConfiguration];
+        // The IAM bearer token endpoint requires a form-encoded request with API key present.
+        // We might as well set that up now.
+        config.HTTPAdditionalHeaders = @{ @"Content-Type" : @"application/x-www-form-urlencoded" };
+        _iamUrlSession = [NSURLSession sessionWithConfiguration:config];
+        [super setShouldMakeSessionRequest:YES];
+        _iamTokenUrl = [NSURL URLWithString:@"https://iam.bluemix.net/oidc/token"];
+    }
+    return self;
+}
+
+- (NSURLSessionConfiguration*)customiseSessionConfig:(NSURLSessionConfiguration*)config
+{
+    // set header for cookie session
+    // The _iam_session endpoint requires an application/json body.
+    // We might as well set that up now.
+    [config setHTTPAdditionalHeaders:@{ @"Content-Type" : @"application/json" }];
+    return config;
+}
+
+- (CDTHTTPInterceptorContext *)interceptRequestInContext:(CDTHTTPInterceptorContext *)context
+{
+    if (self.shouldMakeSessionRequest) {
+        if (!self.cookie) {
+            // We don't have a cookie - first get the IAM bearer token
+            NSData *bearerToken = [self getBearerToken];
+            if (bearerToken != nil) {
+                [self setSessionRequestBody:bearerToken];
+                NSURLComponents *components =
+                [NSURLComponents componentsWithURL:context.request.URL resolvingAgainstBaseURL:NO];
+                components.path = @"/_iam_session";
+                NSURL *url = [components URL];
+                self.cookie = [super startNewSessionAtURL:url withBody:self.sessionRequestBody session:self.urlSession sessionStartedHandler:^(NSData * data){return [self hasSessionStarted:data];}];
+            }
+        }
+    }
+    [context.request setValue:self.cookie forHTTPHeaderField:@"Cookie"];
+    return context;
+}
+
+- (NSData *)getBearerToken {
+
+    CDTLogDebug(CDTREPLICATION_LOG_CONTEXT, @"Getting bearer token");
+    
+    // TODO allow over-riding of URL
```

Comment:
Java: `com.cloudant.client.iamserver`
Python: `IAM_TOKEN_URL`

so maybe `CDT_IAM_TOKEN_URL` to stay close to the Python form?